### PR TITLE
Machine friendly exceptions

### LINKED ISSUE #4: Machine friendly exceptions
JHipster Spring integrates Zalando problems (more precisely the spring version of the lib) to transform Java exception into machine readable exceptions (ie. HTTP status code + a JSON document).

Quarkus does not provide out of the box such feature. The objective is to emulate the same behavior (or simply try to integrate the vanilla Zalando problem lib) in order to be pluggable to the default front applications.

### PR DESCRIPTION
**Pull request by lwitkowski**
Fix #4

### CODE REVIEW

**Comment by danielpetisme** [comment]
That's so awesome!
Regarding tests, you're right we don't have any tests to check the Problem response.
Feel free to add tests to the Java base code if you think it makes sense.

The simpler test you can add is ensuring the files are properly generated
https://github.com/jhipster/generator-jhipster-quarkus/blob/main/test/server.spec.js

File: generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs [hunk 1]
```diff
@@ -18,9 +18,13 @@
 -%>
 package <%=packageName%>.web.rest.errors;
 
+import java.net.URI;
+
 public class EmailAlreadyUsedException extends BadRequestAlertException {
 
+    private static final URI TYPE = URI.create(ErrorConstants.PROBLEM_BASE_URL + "/email-already-used");
```
Comment (danielpetisme) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs:25]: Can you move `TYPE` to `ErrorConstants`, please

File: generators/server/templates/quarkus/src/main/java/package/web/rest/errors/InvalidPasswordWebException.java.ejs [hunk 2]
```diff
@@ -18,19 +18,13 @@
...
-import static javax.ws.rs.core.Response.Status.UNAUTHORIZED;
+import java.net.URI;
 
-import javax.ws.rs.WebApplicationException;
-import javax.ws.rs.core.Response;
+public class InvalidPasswordWebException extends BadRequestAlertException {
 
-public class UserNotAuthenticatedException extends WebApplicationException {
-    private static final long serialVersionUID = 1L;
+    public static final URI TYPE = URI.create(ErrorConstants.PROBLEM_BASE_URL + "/invalid-password");
```
Comment (danielpetisme) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/errors/InvalidPasswordWebException.java.ejs:25]: Same here

File: generators/server/templates/quarkus/src/main/java/package/web/rest/errors/LoginAlreadyUsedException.java.ejs [hunk 3]
```diff
@@ -18,9 +18,13 @@
 -%>
 package <%=packageName%>.web.rest.errors;
 
+import java.net.URI;
+
 public class LoginAlreadyUsedException extends BadRequestAlertException {
 
+    public static final URI TYPE = URI.create(ErrorConstants.PROBLEM_BASE_URL + "/login-already-used");
```
Comment (danielpetisme) [inline generators/server/templates/quarkus/src/main/java/package/web/rest/errors/LoginAlreadyUsedException.java.ejs:25]: Same

**Review by danielpetisme — CHANGES_REQUESTED** [review]

**Comment by danielpetisme** [comment]
Do you plan to add `ExceptionTranslator` also?
The translator adds some headers to the HTTP response on top of the Problem payload.
Again thank you for your contribution!

File: generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs (hunk 1 above)
Comment (lwitkowski) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs:25]: Sure I can, but let me first explain, why I think it's better this way: those specific exceptions are included on condition (!skipUserManagement) - if they are not included, people will see some strange **unused** constants in `ErrorConstants`. I also don't think they need to be shared, and even if - I'd prefer little duplication over coupling and making this conditional inclusion harder. 
But I may be missing some bigger picture here, so I'll happily move them back to `ErrorConstants` class :)

**Comment by lwitkowski** [comment]
> Do you plan to add `ExceptionTranslator` also?

Actually I don't :) All headers are added in `BadRequestAlertException` class and as far as I can see it just works. I find this solution cleaner than `ExceptionTranslator`.

The thing is Zalando's Problem implementation does not have notion of http header - there's no such thing as ProblemWithHeaders, so by default you may have either `Problem`, or customer headers (by using `WebApplicationException`). 

`BadRequestAlertException` is kind of a walk around to have both `application/problem+json` body and custom headers, but it's not generic solution for all exceptions.

Does that make sense?

**Comment by danielpetisme** [comment]
Ok make sense now, smart analysis.

File: generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs (hunk 1 above)
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/errors/EmailAlreadyUsedException.java.ejs:25]: 🤔 I tend to try to keep consistency between JHipster Spring and Quarkus that's why I would recommend using `ErrorConstants`. But I understand your argument...
Let it as is.

**Comment by lwitkowski** [comment]
Thanks for first review @danielpetisme, I'll then focus on adding some tests, and will then mark this PR as 'ready for review'.

**Comment by lwitkowski** [comment]
OK, I added some tests, but not everything is super clear for me which exceptions should have additional headers and which not necessarily, but I guess it's not critical - this version is closer to 'reference' SpringBoot version for sure :)

One more note: `quarkus-resteasy-problem` extension logs all exceptions/problems that flows through it's exception mappers by default, and it looks like this:
`2021-04-08 21:18:07,825 INFO  [http-problem] (executor-thread-1) status=404, title="Not Found", detail="HTTP 404 Not Found"`
We can easily silence them by setting `quarkus.log.category.http-problem.level` to ERROR or OFF - not sure which default is better for this project.

File: generators/server/templates/quarkus/src/test/java/package/web/rest/UserResourceTest.java.ejs [hunk 4]
```diff
@@ -162,7 +166,15 @@ public class UserResourceTest {
             .when()
             .post("/api/users")
             .then()
-            .statusCode(BAD_REQUEST.getStatusCode());
+            .statusCode(BAD_REQUEST.getStatusCode())
+            .header("X-" + applicationName + "-error", is("error.idexists"))
+            .header("X-" + applicationName + "-params", is("userManagement"))
+            .contentType("application/problem+json")
+            .body("title", is("A new user cannot already have an ID"))
+            .body("entityName", is("userManagement"))
```
Comment (lwitkowski) [inline generators/server/templates/quarkus/src/test/java/package/web/rest/UserResourceTest.java.ejs:174]: I'm also not super sure if all those fields (entityName, errorKey, message, params) are needed/used by frontend - I used springboot version as reference and added everything I could find there.

Comment (danielpetisme) [reply generators/server/templates/quarkus/src/test/java/package/web/rest/UserResourceTest.java.ejs:174]: They do

**Comment by danielpetisme** [comment]
> OK, I added some tests, but not everything is super clear for me which exceptions should have additional headers and which not necessarily, but I guess it's not critical - this version is closer to 'reference' SpringBoot version for sure :)
> 
> One more note: `quarkus-resteasy-problem` extension logs all exceptions/problems that flows through it's exception mappers by default, and it looks like this:
> `2021-04-08 21:18:07,825 INFO [http-problem] (executor-thread-1) status=404, title="Not Found", detail="HTTP 404 Not Found"`
> We can easily silence them by setting `quarkus.log.category.http-problem.level` to ERROR or OFF - not sure which default is better for this project.

I propose to let it as is for now.

File: generators/server/templates/quarkus/src/main/java/package/web/util/ResponseUtil.java.ejs [hunk 5]
```diff
@@ -18,21 +18,25 @@
...
-import java.util.Collections;
 import java.util.Map;
 import java.util.Optional;
+import javax.ws.rs.NotFoundException;
 import javax.ws.rs.core.Response;
 
 public interface ResponseUtil {
-    static Response wrapOrNotFound(Optional maybeResponse) {
-        return wrapOrNotFound(maybeResponse, Collections.emptyMap());
+    static Response wrapOrNotFound(Optional<?> maybeResponse) {
```
[the remark this answers is not in the document]
Comment (lwitkowski) [reply generators/server/templates/quarkus/src/main/java/package/web/util/ResponseUtil.java.ejs:27]: You're right, I'll try to minimise the changes to limit the drift from reference spring impl.

File: generators/server/templates/quarkus/src/test/java/package/web/rest/AccountResourceTest_jwt.java.ejs [hunk 6]
```diff
@@ -292,7 +300,20 @@ public class AccountResourceTest {
             .when()
             .post("/api/register")
             .then()
-            .statusCode(BAD_REQUEST.getStatusCode());
+            .statusCode(BAD_REQUEST.getStatusCode())
+            .contentType("application/problem+json")
+            .body("violations", hasSize(2))
+            .body("violations.field", hasItems("email", "password"))
+            .body("violations.error", hasItems("must be a well-formed email address", "size must be between 4 and 100"));
+            // Springboot version returns a bit different schema. Does frontend care/use/display fieldErrors?
```
[the remark this answers is not in the document]
Comment (lwitkowski) [reply generators/server/templates/quarkus/src/test/java/package/web/rest/AccountResourceTest_jwt.java.ejs:308]: Spring version is not perfect (camelCase vs snakcase), `violations` is more in line with zalando problem violations library impl, but I'll add adjust to it, and maybe propose change to reference project first :)

File: generators/server/templates/quarkus/src/main/java/package/web/util/ResponseUtil.java.ejs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lwitkowski) [reply generators/server/templates/quarkus/src/main/java/package/web/util/ResponseUtil.java.ejs:28]: Fixed.

**Comment by lwitkowski** [comment]
@danielpetisme I've addressed all comments (hopefully :), I also updated quarkus-resteasy-problem extension to the latest stable version (which e.g supports quarkus-reactive now - could be relevant for future of this project).

**Review by DanielFran — CHANGES_REQUESTED** [review]

**Review by danielpetisme — APPROVED** [review]

**Comment by danielpetisme** [comment]
LGTM, @DanielFran do you see anything else to update?

**Comment by DanielFran** [comment]
@danielpetisme No, seems OK for me.

**Review by DanielFran — APPROVED** [review]

**Comment by danielpetisme** [comment]
Thank you so much @lwitkowski !
Please, claim the associated bounty
https://www.jhipster.tech/bug-bounties/

**Comment by DanielFran** [comment]
Congratulation @lwitkowski for the contribution, well done.

**Comment by lwitkowski** [comment]
That's awesome, thanks for help @danielpetisme and @DanielFran. I hope it's not my last contribution :)

**Comment by danielpetisme** [comment]
We hope so too 🙂

**Comment by avdev4j** [comment]
thank you @lwitkowski !

I will be happy to see you again :D

**Comment by lwitkowski** [comment]
Here's bounty claim: https://opencollective.com/generator-jhipster/expenses/38542

**Pull request merged**